micro_epsilon_scancontrol_driver/src/driver.cpp

```diff
@@ -550,6 +564,100 @@ namespace scancontrol_driver
         // return true;
     }  
 
+    // Generic function for setting a times like exposure and idle time.
+    // setting_id: 
+    // value: time in microseconds. min: 1 mus, max: 40950 mus ;  eg. 1005 = 1.005ms
+    int ScanControlDriver::SetTime(unsigned int setting_id, unsigned int value){
+        int ret_code;
+
+        // encoded value in 1 mus steps. 
+        uint32_t remainder = ((value % 10) << 12) & 0xF000; // Remainder is left shifted first and bits 0-12 are masked. the quotient will occupy this area.
+        uint32_t quotient = ((value / 10)) & 0xFFF; // take the quotient and mask bits 12-15
+        uint32_t value_encoded = remainder + quotient; 
+
+        ret_code = SetFeature(setting_id, value_encoded); 
+        
+        // success if value was sent AND set.
+        if (ret_code < GENERAL_FUNCTION_OK){
+            RCLCPP_INFO_STREAM(LOGGER, "SetFeature failed.");
+            return ret_code;
+        }
+        
+        unsigned int actual_value = 0;
+        ret_code = GetTime(setting_id, &actual_value);
+        if (actual_value != value){
+            RCLCPP_INFO(LOGGER, "Requested value and actual value do not match. ");
+            return ret_code;
+        }
+        return GENERAL_FUNCTION_OK;
+    }
+
+    int ScanControlDriver::GetTime(unsigned int setting_id, unsigned int* value){
+        int ret_code = 0;
+        ret_code = GetFeature(setting_id, value);
+        
+        if (ret_code < GENERAL_FUNCTION_OK){
+            return ret_code;
+        }
+
+        // Decode
+        uint32_t quotient = *value & 0xFFF;               // Extract ExposureTime / 10
+        uint32_t remainder = (*value >> 12) & 0xF;        // Extract ExposureTime % 10
+        *value = (quotient * 10) + remainder;  // Reconstruct ExposureTime
+
+        return GENERAL_FUNCTION_OK;
+
+    }
+
+
+    // a wrapper on setfeature to use proper encoding 
+    void ScanControlDriver::ServiceSetExposureTime(
+        const std::shared_ptr<micro_epsilon_scancontrol_msgs::srv::SetTime::Request> request,
+        std::shared_ptr<micro_epsilon_scancontrol_msgs::srv::SetTime::Response> response){
+        
+        int ret_code = SetTime(FEATURE_FUNCTION_EXPOSURE_TIME, request->time);
+
+        if (ret_code < GENERAL_FUNCTION_OK){
+            RCLCPP_INFO_STREAM(LOGGER, "Setting exposure time failed with error code: "<<ret_code);
```

Comment:
Change to error severity, also there is double logging now as SetTime also logs.

Reply:
Changed